controllers/keystoneapi_controller.go

```diff
@@ -250,6 +255,72 @@ func (r *KeystoneAPIReconciler) reconcileInit(ctx context.Context, instance *key
 
 	// run keystone db sync - end
 
+	//
+	// expose the service (create service, route and return the created endpoint URLs)
+	//
+	var keystonePorts = map[string]int32{
+		"admin":    keystone.KeystoneAdminPort,
+		"public":   keystone.KeystonePublicPort,
+		"internal": keystone.KeystoneInternalPort,
+	}
+
+	apiEndpoints, ctrlResult, err := common.ExposeEndpoints(
+		ctx,
+		helper,
+		keystone.ServiceName,
+		endpointLabels,
+		keystonePorts,
+	)
+	if err != nil {
+		return ctrlResult, err
+	} else if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+
+	//
+	// Update instance status with service endpoint url from route host information
+	//
+	// TODO: need to support https default here
+	if instance.Status.APIEndpoints == nil {
+		instance.Status.APIEndpoints = map[string]string{}
+	}
+	instance.Status.APIEndpoints = apiEndpoints
+
+	// expose service - end
+
+	//
+	// BootStrap Job
+	//
+	jobDef = keystone.BootstrapJob(instance, instance.Status.APIEndpoints)
+	bootstrapjob := common.NewJob(
+		jobDef,
+		keystonev1.BootstrapHash,
+		instance.Spec.PreserveJobs,
+		5,
+		instance.Status.Hash[keystonev1.BootstrapHash],
+	)
+	ctrlResult, err = bootstrapjob.DoJob(
+		ctx,
+		helper,
+	)
+	if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if bootstrapjob.HasChanged() {
+		instance.Status.Hash[keystonev1.BootstrapHash] = bootstrapjob.GetHash()
+		if err := r.Client.Status().Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.Log.Info(fmt.Sprintf("Job %s hash added - %s", jobDef.Name, instance.Status.Hash[keystonev1.BootstrapHash]))
```

Comment:
You may have explained this before, but is there a reason that we don't rely on the `defer` status-update approach here and instead directly update it in-flight?

Reply:
In general yes we could, but since the hashes are important information to not rerun a job I wanted to save it early when the hash is available and fail early if for some reason we can not update the object before proceed with further steps.